framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -185,13 +190,80 @@ private AnnotationMirror createNumberAnnotationMirror(List<Number> values) {
     }
 
     private TransferResult<CFValue, CFStore> createNewResultBoolean(
-            TransferResult<CFValue, CFStore> result, List<Boolean> resultValues) {
-        AnnotationMirror boolVal =
-                ((ValueAnnotatedTypeFactory) atypefactory).createBooleanAnnotation(resultValues);
-        CFValue newResultValue =
-                analysis.createSingleAnnotationValue(
-                        boolVal, result.getResultValue().getUnderlyingType());
-        return new RegularTransferResult<>(newResultValue, result.getRegularStore());
+            CFStore thenStore,
+            CFStore elseStore,
+            List<Boolean> resultValues,
+            TypeMirror underlyingType) {
+        AnnotationMirror boolVal = atypefactory.createBooleanAnnotation(resultValues);
+        CFValue newResultValue = analysis.createSingleAnnotationValue(boolVal, underlyingType);
+        if (elseStore != null) {
+            return new ConditionalTransferResult<>(newResultValue, thenStore, elseStore);
+        } else {
+            return new RegularTransferResult<>(newResultValue, thenStore);
+        }
+    }
+
+    @Override
+    public TransferResult<CFValue, CFStore> visitFieldAccess(
+            FieldAccessNode node, TransferInput<CFValue, CFStore> in) {
+
+        TransferResult<CFValue, CFStore> result = super.visitFieldAccess(node, in);
+
+        modifyArrayLenBasedOnIntValOnArrayLength(node, result.getRegularStore());
+
+        return result;
+    }
+
+    private void modifyArrayLenBasedOnIntValOnArrayLength(
```

Comment:
This name is unclear.  I would change it to something like "refineArrayWithConstantLength"  or "refineArrayAtLengthAccess".